### PR TITLE
[Main Course] - Recursion

### PR DESCRIPTION
**Pull request by jz263**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] This change requires a documentation update

# How Has This Been Tested?

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: docs/Main_Course/Data_Analytics/2021-08-20-practical-recursion-for-data-cleaning.md [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+    'info': {
+        'governor': 'John Kasich'
+    },
+    'counties': [
+        {'name': 'Summit', 'population': 1234},
+        {'name': 'Cuyahoga', 'population': 1337}]
+    }]
+```
+
+这种嵌套形式的数据如果很复杂，确实不适合人类阅读，如果要对数据进行进一步的清洗或者分析，我们要做的第一步是把嵌套"打开"，转化成类似如下形式的表格：
```
Comment (Realvincentyuan) [inline docs/Main_Course/Data_Analytics/2021-08-20-practical-recursion-for-data-cleaning.md:38]: 进一步的 - 》 `进一步地`

File: docs/Main_Course/Data_Analytics/2021-08-20-practical-recursion-for-data-cleaning.md [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+        if n == 0: # 边界条件
+            return 0
+        elif n == 1: # 边界条件
+            return 1
+        else:
+            return _recurse(n-1) + _recurse(n-2) # 调用自身，缩小问题规模
+    return _recurse(n)
+```
+
+想要求第n个斐波那契数f(n)，我们只需要计算f(n-1)和f(n-2)的值，这一步相当于把问题规模缩小。这个问题可以递推到计算f(2)=f(1)+f(0)。又因为f(1)和f(0)的值是我们已知的边界条件，我们便可以推导出f(2)值，由此逐步得出f(n)。
```
Comment (Realvincentyuan) [inline docs/Main_Course/Data_Analytics/2021-08-20-practical-recursion-for-data-cleaning.md:104]: 可以稍微总结一下递归的重要要素，如：

- 终止条件
- 函数运行状态，每次运行都逐步逼近终止条件
- 调用函数自身

我之前学习递归对的时候看到这些总结会觉得很精辟、会让人印象更加深刻

File: docs/Main_Course/Data_Analytics/2021-08-20-practical-recursion-for-data-cleaning.md [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+
+类似地，在解析嵌套的json数据时，我们可以设计一个函数f()解析当前层的字典，当字典中的某个值(value)为的type为字典或者字典列表时，调用同样的函数f(n)，直到满足value中不含字典这一边界条件。同时，我们需要记录解析遍历的键值对，字典的键即为输出表格中的字段，字典的值为表格中的record。
```
Comment (Realvincentyuan) [inline docs/Main_Course/Data_Analytics/2021-08-20-practical-recursion-for-data-cleaning.md:106]: json -> `JSON`

File: docs/Main_Course/Data_Analytics/2021-08-20-practical-recursion-for-data-cleaning.md [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+| name   | shortname | info    | counties  | name" |
+|--------|-----------|---------|-----------|-------|
+| Florida | FL | {'governor': 'Rick Scott'}  | [{'name': 'Dade', 'population': 12345}, {'name... | NaN   |
+| NaN     | OH | {'governor': 'John Kasich'} | [{'name': 'Summit', 'population': 1234}, {'nam... | Ohio  |
+
+发现pandas只解析了第一层state-level的记录，而county-level的数据还是以嵌套的形式展现。经过亿点点查找，发现pandas自带一个叫做json_normalize的函数可以实现我们的需求:
+
+```Python
+# 使用json_normalize():
+pd.json_normalize(
```
Comment (Realvincentyuan) [inline docs/Main_Course/Data_Analytics/2021-08-20-practical-recursion-for-data-cleaning.md:67]: 可以注明一下pandas的版本，我发现老版本的API调用方式有区别

File: docs/Main_Course/Data_Analytics/2021-08-20-practical-recursion-for-data-cleaning.md [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+
+想要求第n个斐波那契数f(n)，我们只需要计算f(n-1)和f(n-2)的值，这一步相当于把问题规模缩小。这个问题可以递推到计算f(2)=f(1)+f(0)。又因为f(1)和f(0)的值是我们已知的边界条件，我们便可以推导出f(2)值，由此逐步得出f(n)。
+
+类似地，在解析嵌套的json数据时，我们可以设计一个函数f()解析当前层的字典，当字典中的某个值(value)为的type为字典或者字典列表时，调用同样的函数f(n)，直到满足value中不含字典这一边界条件。同时，我们需要记录解析遍历的键值对，字典的键即为输出表格中的字段，字典的值为表格中的record。
+
+我们来看下panads源码中_json_normalize()的核心代码：
+
+``` Python
+
+ _meta = [m if isinstance(m, list) else [m] for m in meta] # 需要展现在结果里的字段
```
Comment (Realvincentyuan) [inline docs/Main_Course/Data_Analytics/2021-08-20-practical-recursion-for-data-cleaning.md:112]: 我看了下源码，是有点长，建议截取的时候把头尾还是加上，然后可以用省略···代替非核心代码，现在的截取方式有点不太对劲，比如最后的`_recursive_extract(data, record_path, {}, level=0)`其实是`_json_normalize()`调用的，加些头部和尾部能避免引起缩进的歧义

File: docs/Main_Course/Data_Analytics/2021-08-20-practical-recursion-for-data-cleaning.md [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+            records.extend(recs) 
+
+_recursive_extract(data, record_path, {}, level=0)
+```
+
+可以看出源码还是稍微有些难以理解的，此外代码作者也吐槽目前的执行效率很低，但这个函数好在可以通过定义record_path控制解析的深度，防止将过深的半结构化数据完全展开，产生过多的记录。
+
+## 3. 总结
+
+虽然数据分析师的日常工作中往往不需要直接接触算法，但掌握一些算法思想往往可以帮助我们更好地理解手边的工具。必要时也可以自己编写代码提高效率。除了刷算法题之外，读读优秀开源项目地源码也是很好的提升方式。如果实在没有时间，那就常来我们的公众号上看看吧！
```
Comment (Realvincentyuan) [inline docs/Main_Course/Data_Analytics/2021-08-20-practical-recursion-for-data-cleaning.md:173]: lol

**Review by Realvincentyuan — APPROVED** [review]
有些小评论

File: docs/Main_Course/Data_Analytics/2021-08-20-practical-recursion-for-data-cleaning.md [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+---
+
+# 利用递归思想处理半结构化数据
+
+!!! info
+    作者：Jermey，发布于2021-08-22，阅读时间：约6分钟，微信公众号文章链接：[:fontawesome-solid-link:]
+
+## 1. 背景
+
+在日常数据分析的工作中，我们收集到的原始数据有时并不是整齐的表格形式，例如在爬取网页或者爬取api里的数据时，结果往往是以xml或者json（类似Python中的字典）格式返回，并且层层嵌套，就像如下这样：
```
Comment (jhqian0909) [inline docs/Main_Course/Data_Analytics/2021-08-20-practical-recursion-for-data-cleaning.md:12]: 可以强调下如下的是json格式的

File: docs/Main_Course/Data_Analytics/2021-08-20-practical-recursion-for-data-cleaning.md (hunk 1 above)
Comment (jz263) [reply docs/Main_Course/Data_Analytics/2021-08-20-practical-recursion-for-data-cleaning.md:38]: 已修改

File: docs/Main_Course/Data_Analytics/2021-08-20-practical-recursion-for-data-cleaning.md (hunk 2 above)
Comment (jz263) [reply docs/Main_Course/Data_Analytics/2021-08-20-practical-recursion-for-data-cleaning.md:104]: 确实，已添加

File: docs/Main_Course/Data_Analytics/2021-08-20-practical-recursion-for-data-cleaning.md (hunk 3 above)
Comment (jz263) [reply docs/Main_Course/Data_Analytics/2021-08-20-practical-recursion-for-data-cleaning.md:106]: 已修改

File: docs/Main_Course/Data_Analytics/2021-08-20-practical-recursion-for-data-cleaning.md (hunk 4 above)
Comment (jz263) [reply docs/Main_Course/Data_Analytics/2021-08-20-practical-recursion-for-data-cleaning.md:67]: 已注明

File: docs/Main_Course/Data_Analytics/2021-08-20-practical-recursion-for-data-cleaning.md (hunk 5 above)
Comment (jz263) [reply docs/Main_Course/Data_Analytics/2021-08-20-practical-recursion-for-data-cleaning.md:112]: 已添加头尾部代码

File: docs/Main_Course/Data_Analytics/2021-08-20-practical-recursion-for-data-cleaning.md (hunk 7 above)
Comment (jz263) [reply docs/Main_Course/Data_Analytics/2021-08-20-practical-recursion-for-data-cleaning.md:12]: 已修改

**Pull request merged**